src/java.base/share/classes/jdk/internal/foreign/layout/AbstractGroupLayout.java

```diff
@@ -0,0 +1,166 @@
+import java.util.Optional;
+import java.util.function.LongBinaryOperator;
+import java.util.stream.Collectors;
+
+/**
+ * A compound layout that aggregates multiple <em>member layouts</em>. There are two ways in which member layouts
+ * can be combined: if member layouts are laid out one after the other, the resulting group layout is said to be a <em>struct</em>
+ * (see {@link MemoryLayout#structLayout(MemoryLayout...)}); conversely, if all member layouts are laid out at the same starting offset,
+ * the resulting group layout is said to be a <em>union</em> (see {@link MemoryLayout#unionLayout(MemoryLayout...)}).
+ *
+ * @implSpec
+ * This class is immutable, thread-safe and <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>.
+ *
+ * @since 19
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+public abstract class AbstractGroupLayout<L extends AbstractGroupLayout<L> & MemoryLayout> extends AbstractLayout<L> {
+
+    /**
+     * The group kind.
+     */
+    enum Kind {
+        /**
+         * A 'struct' kind.
+         */
+        STRUCT("", Math::addExact),
+        /**
+         * A 'union' kind.
+         */
+        UNION("|", Math::max);
+
+        final String delimTag;
+        final LongBinaryOperator sizeOp;
+
+        Kind(String delimTag, LongBinaryOperator sizeOp) {
+            this.delimTag = delimTag;
+            this.sizeOp = sizeOp;
+        }
+
+        long sizeof(List<MemoryLayout> elems) {
+            long size = 0;
+            for (MemoryLayout elem : elems) {
+                size = sizeOp.applyAsLong(size, elem.bitSize());
+            }
+            return size;
+        }
+
+        long alignof(List<MemoryLayout> elems) {
+            return elems.stream().mapToLong(MemoryLayout::bitAlignment).max() // max alignment in case we have member layouts
+                    .orElse(1); // or minimal alignment if no member layout is given
+        }
+    }
+
+    private final Kind kind;
+    private final List<MemoryLayout> elements;
+
+    AbstractGroupLayout(Kind kind, List<MemoryLayout> elements) {
+        this(kind, elements, kind.alignof(elements), Optional.empty());
+    }
+
+    AbstractGroupLayout(Kind kind, List<MemoryLayout> elements, long alignment, Optional<String> name) {
+        super(kind.sizeof(elements), alignment, name);
+        this.kind = kind;
+        this.elements = List.copyOf(elements);
+    }
+
+    /**
+     * Returns the member layouts associated with this group.
+     *
+     * @apiNote the order in which member layouts are returned is the same order in which member layouts have
+     * been passed to one of the group layout factory methods (see {@link MemoryLayout#structLayout(MemoryLayout...)},
+     * {@link MemoryLayout#unionLayout(MemoryLayout...)}).
+     *
+     * @return the member layouts associated with this group.
+     */
+    public List<MemoryLayout> memberLayouts() {
+        return Collections.unmodifiableList(elements);
```

Comment:
`elements` should already be unmodifable. The constructor performs `List.copyOf()`, and that's a noop if the list passed was produced from that method or `List.of`.